### PR TITLE
add steps

### PR DESCRIPTION
**Pull request by EmilHvitfeldt**
This PR sets up 3 recipe steps with dummy names for use in feature selection. These have been set up to follow the standards that we use in the rest of the tidymodels recipes packages

### CODE REVIEW

**Comment by franceslinyc** [comment]
I literally didn't do anything yet except filling in the properties/attributes. Working on filtro's latest PR, just FYI!

**Comment by topepo** [comment]
@jrosell I was about to hit the button to send this to CRAN and saw this. 🙀 Thanks for posting this. 

I had a few thoughts. I was surprised to see the error go down as the noise predictors were retained. The amount is not much, reaching the MAE levels that we see with no noise predictors, and the overall difference in the delivery time MAE is about _nine seconds_. The signal is clear but tiny. 

This particular data set has a handful of very important predictors and about 35ish `item` predictors that have a negligible effect on the outcome, especially without a few interactions. Plain old linear regression isn’t robust against extra predictors (like glmnet), but it isn’t particularly sensitive to them in “small” amounts. This is especially true if they are not very correlated. However, this _should_ result in a relatively flat pattern in MAE, not an increasing one. 

I can verify that the recipe step ranks the predictors appropriately, and the data being held out is not involved in the selection, so data leakage probably isn’t the issue. The same data is used to select predictors and model them. That’s not optimal (compared to full-on RFE). However, if that were the issue, we would be overfitting the predictor set, and the assessment data would show poor results. 

One thing that I would do differently is to apply the filter first in the recipe. That eliminates a lot more terms at once. Correlating spline terms to the outcome might not work well since they are designed to model a local area in predictor space. That might dilute the correlation. 

I did modify the recipe to see if that was the issue, and I got the same result pattern. 

So again, there is surprise, but it’s not a big effect. 

So what would happen if the [model were _very_ sensitive](https://aml4td.org/chapters/feature-selection.html#fig-irrelevant-predictors) to extra noise columns? I used the same neural network tuned in the “whole game” chapter and used 24 units (same as the book). I ran the filter first, then did the existing feature engineering (code below).

Here are the results: 

image

If there are about 6-10 predictors, that is a proportion range of 0.1 to 0.17. The best point is around 0.13 and shows underfitting before that, and the effect of too many predictors after. 

(The weird spikes are cases when the gradient search went off the rails, causing a numeric overflow in the loss function. I’m still working on fixing that.)

This result is consistent with what I would expect, so I feel comfortable that it is working as intended. 

One other thing, the desirability2 package can let you make tradeoffs between performance and the amount of feature selection via

```r
library(desirability2)
tune_res |> 
	show_best_desirability(
	minimize(mae), 
	minimize(prop_terms, scale = 2)
)
```

so that can help make a good choice. 

Other code: 

```r
norm_rec <- 
	recipe(time_to_delivery ~ ., data = noise_train) %>% 
	step_predictor_best(
		all_predictors(), 
		prop_terms = tune(),
		score = "cor_pearson",
	) |> 
	step_dummy(all_factor_predictors()) %>% 
	step_zv(all_predictors()) %>% 
	step_normalize(all_predictors())

nnet_spec <- 
	mlp(
		hidden_units = 24,
		penalty = 0.01,
		learn_rate = 0.1,
		epochs = 5000
	) %>%
	set_mode("regression") %>%
	set_engine("brulee", stop_iter = 10, rate_schedule = "cyclic")

nnet_wflow <- 
	workflow() %>% 
	add_model(nnet_spec) %>% 
	add_recipe(norm_rec)

nnet_grid <- tibble(prop_terms = seq(0.05, 1, length.out = 50))

set.seed(388)
nnet_res <-
	tune_grid(nnet_wflow,
						noise_folds,
						grid = nnet_grid,
						metrics = metric_set(mae, rsq))

nnet_res |>
	collect_metrics(summarize = FALSE) |>
	summarize(
		mean = mean(.estimate),
		median = median(.estimate),
		.by = c(prop_terms, .metric)
	) |>
	pivot_longer(
		cols = c(mean, median),
		names_to = "estimate",
		values_to = "value"
	) |>
	ggplot(aes(prop_terms, value, col = estimate, group = estimate)) + 
	geom_point(cex = 1) + 
	geom_line() + 
	facet_wrap(~ .metric, ncol = 1, scales = "free_y") + 
	geom_vline(xintercept = c(0.1, 0.17), lty = 3) + 
	theme_bw()
```
